src/database/UIDatabase.js

```diff
@@ -198,4 +198,22 @@ export const getUIDatabaseInstance = database => {
   return UIDatabaseInstance;
 };
 
+// Factory function for custom UIDatabase PropTypes.
+const createUIDatabasePropType = isRequired => (props, propName, componentName) => {
+  const { [propName]: prop } = props;
+
+  if (prop == null) {
+    if (isRequired) {
+      return new TypeError(`Missing database prop in ${componentName}`);
+    }
+    return null;
+  }
+
+  if (prop instanceof UIDatabase) return null;
+  return new TypeError(`Invalid database prop in ${componentName}`);
+};
+
+export const UIDatabaseType = createUIDatabasePropType(false);
+UIDatabaseType.isRequired = createUIDatabasePropType(true);
+
```

Comment:
this is cool is it a common pattern I've never seen it before

Reply:
I couldn't find any examples with a similar "singleton" design to ours, so this is mostly just a pattern I came up with that I thought went well with our singleton design. But the `.isRequired` is just a property like any other (actually just a convention), so I saw most common prop types using some sort of closure or nested function call in a similar way.